src/WebJobs.Extensions.DurableTask/EventGridLifeCycleNotificationHelper.cs

```
@@ -42,91 +42,123 @@ public EventGridLifeCycleNotificationHelper(
+                this.eventGridTopicEndpoint = nameResolver.Resolve("EventGrid:topicEndpoint");
+
+                if (string.Equals(nameResolver.Resolve("EventGrid:credential"), "managedidentity") &&
+                    !string.IsNullOrEmpty(nameResolver.Resolve("EventGrid:clientId")))
+                {
+                    // Use user assigned managed identity
+                    string clientId = nameResolver.Resolve("EventGrid:clientId");
+                    this.ManagedIdentityTokenSource = new ManagedIdentityTokenSource("https://eventgrid.azure.net/.default", new ManagedIdentityOptions { ClientId = clientId });
+                }
+                else
+                {
+                    // Use system assigned managed identity
+                    this.ManagedIdentityTokenSource = new ManagedIdentityTokenSource("https://eventgrid.azure.net/.default");
+                }
+            }
+            else
+            {
+                eventGridNotificationsConfig = options.Notifications.EventGrid ?? throw new ArgumentNullException(nameof(options.Notifications.EventGrid));
+
+                this.eventGridKeyValue = nameResolver.Resolve(eventGridNotificationsConfig.KeySettingName);
+
+                this.eventGridTopicEndpoint = eventGridNotificationsConfig.TopicEndpoint;
+
+                if (nameResolver.TryResolveWholeString(eventGridNotificationsConfig.TopicEndpoint, out var endpoint))
+                {
+                    this.eventGridTopicEndpoint = endpoint;
+                }
+            }
+
+            bool isTopicEndpointKeyConfigured = !string.IsNullOrEmpty(this.eventGridTopicEndpoint) && eventGridNotificationsConfig != null;
 
-            if (!string.IsNullOrEmpty(this.eventGridTopicEndpoint))
-            {
-                if (!string.IsNullOrEmpty(eventGridNotificationsConfig.KeySettingName))
-                {
-                    this.useTrace = true;
+            if (this.UseManagedIdentity || isTopicEndpointKeyConfigured)
```

Again should we have a check here for if `this.eventGridKeyValue` is non-null too? Like `this.UseManagedIdentity || (isTopicEndpointKeyConfigured && !string.IsNullOrEmpty(this.eventGridKeyValue)` or something along those lines (this is assuming that when not using managed identity the event grid key needs to be specified)

---

I tried to follow the current implementation to keep the structure the same so the exceptions throw at the end after checking for `KeySettingName` and `eventGridValue`. It's probably better to throw the exceptions at the beginning though. I can make that change.